src/combinatorics/simple_graph/connectivity.lean

```diff
@@ -652,42 +652,49 @@ begin
     exact ⟨λ h', h.1 (fst_mem_support_of_mem_edges p_p h'), p_ih h.2⟩, }
 end
 
+variables {u v w x y : V} {p : G.walk v w}
+
+/-- Returns whether a walk is the empty walk. Useful because `p = simple_graph.walk.nil` does not
+typecheck. -/
+inductive is_nil : ∀ {x y : V}, G.walk x y → Prop
+| nil {x : V} : is_nil (nil : G.walk x x)
+
+protected lemma is_nil.eq (hp : p.is_nil) : v = w := by cases hp; refl
+
+@[simp] lemma not_is_nil_of_ne : v ≠ w → ¬ p.is_nil := mt is_nil.eq
+
 /-- The first dart of a path. -/
-@[simp]
-def head : Π {x y : V}, G.walk x y → x ≠ y → G.dart
-| x _ nil         h := (h rfl).elim
-| x z (cons hxz p) h := ⟨⟨x, _⟩, hxz⟩
+def head (p : G.walk x y) (hp : ¬ p.is_nil) : G.dart :=
+{ fst := x,
+  snd := by { cases p, cases hp is_nil.nil, exact p_v },
+  is_adj := by { obtain _ | ⟨h, _⟩ := p, cases hp is_nil.nil, exact h } }
 
 /-- The path obtained by removing the first dart of a path. -/
 @[simp]
-def tail : Π {x y : V} (p : G.walk x y) (hxy : x ≠ y), G.walk (p.head hxy).snd y
-| x _ nil         h := (h rfl).elim
+def tail : Π {x y : V} (p : G.walk x y) (hp : ¬ p.is_nil), G.walk (p.head hp).snd y
+| x _ nil         h := (h is_nil.nil).elim
 | x z (cons hxz p) h := p
 
-@[simp] lemma head_fst : ∀ {x y : V} (p : G.walk x y) (hxy : x ≠ y), (p.head hxy).fst = x
-| _ _ nil h := (h rfl).elim
-| _ _ (cons hxz w) _ := rfl
+@[simp] lemma head_fst {x y : V} (p : G.walk x y) (hp : ¬ p.is_nil) : (p.head hp).fst = x := rfl
 
-private lemma adj_head_snd {x y : V} (p : G.walk x y) (hxy : x ≠ y) : G.adj x (p.head hxy).snd :=
-by { convert  (p.head hxy).is_adj, rw head_fst }
+private lemma adj_head_snd {x y : V} (p : G.walk x y) (hp : ¬ p.is_nil) : G.adj x (p.head hp).snd :=
```

Comment:
I feel like this shouldn't be private, I think it can be useful outside of this file

Reply:
Oh right. I made it private back when `head_fst` wasn't refl. Now it makes sense for it to be part of the API.